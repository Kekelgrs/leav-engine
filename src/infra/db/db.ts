import {Database} from 'arangojs';

interface IDeps {
    config?: any;
}

const _getDbConnection = (config: any): Database => {
    return new Database({
        url: config.db.url
    });
};

export default function(deps: IDeps = {}): Database {
    const db = _getDbConnection(deps.config);

    db.useDatabase(deps.config.db.name);

    return db;
}

export const initDb = async (config: any) => {
    const db = _getDbConnection(config);

    const databases = await db.listDatabases();
    const dbExists = databases.reduce((exists, d) => exists || d === config.db.name, false);

    if (!dbExists) {
        await db.createDatabase(config.db.name);
    }

    db.close();
};
