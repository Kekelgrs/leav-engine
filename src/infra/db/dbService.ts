import {Database} from 'arangojs';

export default function(config: any) {
    const db = new Database({
        url: config.db.url
    });
    db.useDatabase(config.db.name);

    return db;
}
