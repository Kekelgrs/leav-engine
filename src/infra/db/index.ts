export {default} from './db';
export {default as dbService} from './dbService';
export {default as dbUtils} from './dbUtils';
