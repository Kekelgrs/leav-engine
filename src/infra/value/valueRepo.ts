import {IAttribute} from '_types/attribute';
import {IValue, IValuesOptions} from '_types/value';
import {IAttributeTypesRepo} from '../attributeTypes/attributeTypesRepo';

export interface IValueRepo {
    /**
     * Create a new value
     *
     * @param library
     * @param recordId
     * @param attribute
     * @param value
     * @param {}
     */
    createValue(library: string, recordId: number, attribute: IAttribute, value: IValue): Promise<IValue>;

    /**
     * Update an existing value. Field "id" is expected on the value
     * @param library
     * @param recordId
     * @param attribute
     * @param value
     * @param {}
     */
    updateValue(library: string, recordId: number, attribute: IAttribute, value: IValue): Promise<IValue>;

    /**
     * Delete an existing value. Field "id" is expected on the value
     *
     * @param library
     * @param recordId
     * @param attribute
     * @param value
     * @param {}
     */
    deleteValue(library: string, recordId: number, attribute: IAttribute, value: IValue): Promise<IValue>;

    /**
     * Get all values for given record and attribute
     *
     * @param library
     * @param recordId
     * @param attribute
     * @return Array<{}>    Return an empty array if no value found
     */
    getValues(
        library: string,
        recordId: number,
        attribute: IAttribute,
        forceGetAllValues?: boolean,
        options?: IValuesOptions
    ): Promise<IValue[]>;

    /**
     * Return a specific value based on its ID. Field "id" is expect on the value
     *
     * @param library
     * @param recordId
     * @param attribute
     * @param value
     * @return {}   Return null if no value found
     */
    getValueById(library: string, recordId: number, attribute: IAttribute, value: IValue): Promise<IValue>;
    clearAllValues(attribute: IAttribute): Promise<boolean>;
}

interface IDeps {
    'core.infra.attributeTypes'?: IAttributeTypesRepo;
}

export default function({'core.infra.attributeTypes': attributeTypesRepo = null}: IDeps = {}): IValueRepo {
    return {
        createValue(library: string, recordId: number, attribute: IAttribute, value: IValue): Promise<IValue> {
            const typeRepo = attributeTypesRepo.getTypeRepo(attribute);
            return typeRepo.createValue(library, recordId, attribute, value);
        },
        updateValue(library: string, recordId: number, attribute: IAttribute, value: IValue): Promise<IValue> {
            const typeRepo = attributeTypesRepo.getTypeRepo(attribute);
            return typeRepo.updateValue(library, recordId, attribute, value);
        },
        deleteValue(library: string, recordId: number, attribute: IAttribute, value: IValue): Promise<IValue> {
            const typeRepo = attributeTypesRepo.getTypeRepo(attribute);
            return typeRepo.deleteValue(library, recordId, attribute, value);
        },
        getValues(
            library: string,
            recordId: number,
            attribute: IAttribute,
            forceGetAllValues?: boolean,
            options?: IValuesOptions
        ): Promise<IValue[]> {
            const typeRepo = attributeTypesRepo.getTypeRepo(attribute);
            return typeRepo.getValues(library, recordId, attribute, forceGetAllValues, options);
        },
        getValueById(library: string, recordId: number, attribute: IAttribute, value: IValue): Promise<IValue> {
            const typeRepo = attributeTypesRepo.getTypeRepo(attribute);
            return typeRepo.getValueById(library, recordId, attribute, value);
        },
        clearAllValues(attribute: IAttribute): Promise<boolean> {
            const typeRepo = attributeTypesRepo.getTypeRepo(attribute);
            return typeRepo.clearAllValues(attribute);
        }
    };
}
