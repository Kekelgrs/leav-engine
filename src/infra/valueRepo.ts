import {IAttribute, AttributeTypes} from '../domain/attributeDomain';
import {IValue} from '../domain/valueDomain';
import {IDbService} from './db/dbService';

export interface IValueRepo {
    saveValue?(library: string, recordId: string, attribute: IAttribute, value: IValue): Promise<IValue>;
}

export default function(dbService: IDbService | any): IValueRepo {
    return {
        async saveValue(library: string, recordId: string, attribute: IAttribute, value: IValue): Promise<IValue> {
            if (attribute.type === AttributeTypes.INDEX) {
                const collec = dbService.db.collection(library);
                const updatedDoc = await collec.update(
                    {
                        _key: recordId
                    },
                    {
                        [attribute.id]: value.value
                    }
                );

                const docData = await collec.document(updatedDoc);

                return {value: docData[attribute.id]};
            }
        }
    };
}
