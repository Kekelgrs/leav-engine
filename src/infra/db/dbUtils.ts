import {aql} from 'arangojs';
import {GeneratedAqlQuery} from 'arangojs/lib/cjs/aql-query';
import {CollectionType} from 'arangojs/lib/cjs/collection';
import {AwilixContainer} from 'awilix';
import {readdirSync} from 'fs';
import {IPluginsRepo} from 'infra/plugins/pluginsRepo';
import * as path from 'path';
import {isArray} from 'util';
import * as winston from 'winston';
import {IAttribute, IAttributeFilterOptions} from '_types/attribute';
import {IForm} from '_types/forms';
import {ILibrary, ILibraryFilterOptions} from '_types/library';
import {IList, IPaginationParams, ISortParams} from '_types/list';
import {IQueryInfos} from '_types/queryInfos';
import {IKeyValue} from '_types/shared';
import {ITree, ITreeFilterOptions} from '_types/tree';
import {IDbValueVersion, IValueVersion} from '_types/value';
import {collectionTypes, IDbService, IExecuteWithCount} from './dbService';
import runMigrationFiles from './helpers/runMigrationFiles';

export const MIGRATIONS_COLLECTION_NAME = 'core_db_migrations';

export interface IFindCoreEntityParams {
    collectionName: string;
    filters?: ITreeFilterOptions | ILibraryFilterOptions | IAttributeFilterOptions;
    strictFilters?: boolean;
    withCount?: boolean;
    pagination?: IPaginationParams;
    sort?: ISortParams;
    customFilterConditions?: IKeyValue<
        (filterKey: string, filterVal: string | boolean | string[], strictFilters: boolean) => GeneratedAqlQuery
    >;
    ctx: IQueryInfos;
}

export interface IDbUtils {
    migrate?(depsManager: AwilixContainer): Promise<void>;
    cleanup?(record: {}): any;
    convertToDoc?(obj: {}): any;
    isCollectionExists?(name: string): Promise<boolean>;
    findCoreEntity?<T extends ITree | ILibrary | IAttribute | IForm>(params: IFindCoreEntityParams): Promise<IList<T>>;
    convertValueVersionToDb?(version: IValueVersion): IDbValueVersion;
    convertValueVersionFromDb?(version: IDbValueVersion): IValueVersion;
    clearDatabase(): Promise<void>;
}

interface IDeps {
    'core.infra.db.dbService'?: IDbService;
    'core.utils.logger'?: winston.Winston;
    'core.infra.plugins'?: IPluginsRepo;
    config?: any;
}

export default function({
    'core.infra.db.dbService': dbService = null,
    'core.utils.logger': logger = null,
    'core.infra.plugins': pluginsRepo = null,
    config = null
}: IDeps = {}): IDbUtils {
    /**
     * Create the collections used to managed db migrations
     * This can't be in a migration file because we do have to initialize it somewhere
     *
     */
    async function _initMigrationsCollection(): Promise<void> {
        const collections = await dbService.db.listCollections();

        const colExists = collections.reduce((exists, c) => exists || c.name === MIGRATIONS_COLLECTION_NAME, false);
        if (!colExists) {
            const collection = dbService.db.collection(MIGRATIONS_COLLECTION_NAME);
            await collection.create();
        }
    }

    /**
     * Return the filter's conditions based on key and val supplied.
     *
     * @param filterKey
     * @param filterVal
     * @param bindVars
     * @param index
     * @param strictFilters
     */
    function _getFilterCondition(
        filterKey: string,
        filterVal: string | boolean | string[],
        strictFilters: boolean
    ): GeneratedAqlQuery {
        const queryParts = [];

        // If value is an array (types or formats for example),
        // we call this function recursively on array and join filters with an OR
        if (Array.isArray(filterVal)) {
            if (filterVal.length) {
                const valParts = filterVal.map(val => _getFilterCondition(filterKey, val, strictFilters));
                queryParts.push(aql.join(valParts, ' OR '));
            }
        } else {
            if (filterKey === 'label') {
                // Search for label in any language
                const valParts = config.lang.available.map(l => aql`LIKE(el.label.${l}, ${filterVal}, true)`);
                queryParts.push(aql.join(valParts, ' OR '));
            } else {
                // Filter with a "like" on ID or exact value in other fields
                queryParts.push(
                    filterKey === '_key' && !strictFilters
                        ? aql`LIKE(el.${filterKey}, ${filterVal}, true)`
                        : aql`el.${filterKey} == ${filterVal}`
                );
            }
        }

        return aql.join(queryParts);
    }

    const ret = {
        /**
         * Run database migrations.
         * It takes all files present in migrations folder and run it if it's never been executed before
         *
         * @param depsManager
         */
        async migrate(depsManager: AwilixContainer): Promise<void> {
            await _initMigrationsCollection();
            const ctx: IQueryInfos = {
                userId: '0',
                queryId: 'run-migrations'
            };
            // Load already ran migrations
            const executedMigrations = await dbService.execute({
                query: `
                    FOR m IN core_db_migrations
                    RETURN m.file
                `,
                ctx
            });

            const _runMigrationFiles = (files, folder, prefix = null) =>
                runMigrationFiles({
                    files,
                    executedMigrations,
                    migrationsDir: folder,
                    prefix,
                    deps: {depsManager, dbService, logger},
                    ctx
                });

            /*** Core migrations ***/
            // Load migrations files
            const migrationsDir = path.resolve(__dirname, 'migrations');
            const migrationFiles = readdirSync(migrationsDir).filter(file => file.indexOf('.map') === -1);

            await _runMigrationFiles(migrationFiles, migrationsDir);

            /*** Plugins migrations ***/
            const plugins = pluginsRepo.getRegisteredPlugins();
            for (const plugin of plugins) {
                const pluginMigrationFolder = path.resolve(plugin.path + '/infra/db/migrations');
                const pluginMigrationFiles = readdirSync(pluginMigrationFolder).filter(
                    file => file.indexOf('.map') === -1
                );

                await _runMigrationFiles(pluginMigrationFiles, pluginMigrationFolder, plugin.infos.name);
            }
        },

        /**
         * Cleanup every system keys from an object coming from database.
         * _key is kept under 'id'
         *
         * @param obj
         * @return any   Cleaned up object
         */
        cleanup(obj: any): any {
            if (obj === null) {
                return null;
            }

            return Object.keys(obj).reduce((newObj: any, key) => {
                if (key === '_key') {
                    newObj.id = obj[key];
                } else if (key[0] !== '_') {
                    newObj[key] = obj[key];
                }

                return newObj;
            }, {});
        },

        /**
         * Convert an object to an object looking like a DB document
         * id is replaced by _key
         *
         * @param obj
         * @return any   DB document compatible object
         */
        convertToDoc(obj: {}): any {
            const newObj: any = {...obj};

            if (typeof newObj.id !== 'undefined') {
                newObj._key = newObj.id;
            }
            delete newObj.id;

            return newObj;
        },
        /**
         * Search core entities (libraries, attributes, trees)
         *
         * @param collectionName
         * @param filters
         * @param strictFilters
         */
        async findCoreEntity<T extends ITree | ILibrary | IAttribute>(
            params: IFindCoreEntityParams
        ): Promise<IList<T>> {
            const {
                collectionName = null,
                filters = null,
                strictFilters = false,
                withCount = false,
                pagination = null,
                sort = null,
                customFilterConditions = {},
                ctx = {}
            } = params;

            const collec = dbService.db.collection(collectionName);
            const queryParts = [aql`FOR el IN ${collec}`];

            if (filters !== null) {
                const dbFilters = ret.convertToDoc(filters);
                const filtersKeys = Object.keys(dbFilters);

                for (const filterKey of filtersKeys) {
                    const filterVal = dbFilters[filterKey];

                    // Caller can define some custom functions to generate filter condition (like looking on edges to
                    // filter on libraries linked to an attribute). So, if a custom function is define for this filter,
                    // we use it, otherwise we use the standard filters
                    const filterCondsFunc =
                        typeof customFilterConditions[filterKey] !== 'undefined'
                            ? customFilterConditions[filterKey]
                            : _getFilterCondition;
                    const conds = filterCondsFunc(filterKey, filterVal, strictFilters);

                    if (conds.query) {
                        queryParts.push(aql`FILTER`, conds);
                    }
                }
            }

            if (!!sort) {
                const field = sort.field === 'id' ? '_key' : sort.field;
                queryParts.push(aql`SORT el.${field} ${sort.order}`);
            }

            if (!!pagination) {
                queryParts.push(aql`LIMIT ${pagination.offset || 0}, ${pagination.limit}`);
            }

            queryParts.push(aql`RETURN el`);

            const query = aql.join(queryParts);
            const res = await dbService.execute<IExecuteWithCount | any[]>({query, withTotalCount: withCount, ctx});

            const results = !isArray(res) ? res.results : res;

            return {
                totalCount: withCount ? (res as IExecuteWithCount).totalCount : null,
                list: results.map(ret.cleanup)
            };
        },
        convertValueVersionToDb(version: IValueVersion): IDbValueVersion {
            return Object.keys(version).reduce((allVers, treeName) => {
                const {library, id} = version[treeName];

                allVers[treeName] = `${library}/${id}`;

                return allVers;
            }, {});
        },
        convertValueVersionFromDb(version: IDbValueVersion): IValueVersion {
            return version
                ? Object.keys(version).reduce((allVers, treeName) => {
                      const [library, id] = version[treeName].split('/');

                      allVers[treeName] = {
                          library,
                          id
                      };

                      return allVers;
                  }, {})
                : null;
        },
        async clearDatabase(): Promise<void> {
            // Drop all collections
            const cols = await dbService.db.listCollections();
            for (const col of cols) {
                const colType =
                    col.type === CollectionType.DOCUMENT_COLLECTION ? collectionTypes.DOCUMENT : collectionTypes.EDGE;
                await dbService.dropCollection(col.name, colType);
            }
        }
    };

    return ret;
}
