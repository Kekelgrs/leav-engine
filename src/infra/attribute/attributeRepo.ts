import {aql} from 'arangojs';
import {IList, IPaginationParams} from '_types/list';
import {AttributeFormats, AttributeTypes, IAttribute, IAttributeFilterOptions} from '../../_types/attribute';
import {ILibrary} from '../../_types/library';
import {IDbService} from '../db/dbService';
import {IDbUtils} from '../db/dbUtils';
import {LIB_ATTRIB_COLLECTION_NAME} from '../library/libraryRepo';
import {IValueRepo} from '../value/valueRepo';

export interface IAttributeRepo {
    getAttributes(
        filters?: IAttributeFilterOptions,
        strictFilters?: boolean,
        withCount?: boolean,
        pagination?: IPaginationParams
    ): Promise<IList<IAttribute>>;
    updateAttribute(attrData: IAttribute): Promise<IAttribute>;
    createAttribute(attrData: IAttribute): Promise<IAttribute>;
    deleteAttribute(attrData: IAttribute): Promise<IAttribute>;

    /**
     * Get all libraries for which this attribute is enabled
     *
     * @param attribute
     */
    getLibrariesUsingAttribute?(attribute: IAttribute): Promise<ILibrary[]>;
}

export const ATTRIB_COLLECTION_NAME = 'core_attributes';

export default function(
    dbService: IDbService | null = null,
    dbUtils: IDbUtils | null = null,
    valueRepo: IValueRepo | null = null,
    config = null
): IAttributeRepo {
    return {
        async getAttributes(
            filters?: IAttributeFilterOptions,
            strictFilters: boolean = false,
            withCount: boolean = false,
            pagination?: IPaginationParams
        ): Promise<IList<IAttribute>> {
            return dbUtils.findCoreEntity<IAttribute>(
                ATTRIB_COLLECTION_NAME,
                filters,
                strictFilters,
                withCount,
                pagination
            );
        },
        async updateAttribute(attrData: IAttribute): Promise<IAttribute> {
            const defaultParams = {
                _key: '',
                system: false,
                label: {fr: '', en: ''},
                type: AttributeTypes.ADVANCED,
                format: AttributeFormats.TEXT,
                multiple_values: false
            };
            let docToInsert = dbUtils.convertToDoc(attrData);

            docToInsert = {...defaultParams, ...docToInsert};

            // Insert in libraries collection
            const col = dbService.db.collection(ATTRIB_COLLECTION_NAME);
            const res = await dbService.execute(aql`UPDATE ${docToInsert} IN ${col} RETURN NEW`);

            return dbUtils.cleanup(res.pop());
        },
        async createAttribute(attrData: IAttribute): Promise<IAttribute> {
            const defaultParams = {
                _key: '',
                system: false,
                label: {fr: '', en: ''},
                type: AttributeTypes.ADVANCED,
                format: AttributeFormats.TEXT,
                multiple_values: false
            };
            let docToInsert = dbUtils.convertToDoc(attrData);

            docToInsert = {...defaultParams, ...docToInsert};

            // Insert in libraries collection
            const col = dbService.db.collection(ATTRIB_COLLECTION_NAME);
            const res = await dbService.execute(aql`INSERT ${docToInsert} IN ${col} RETURN NEW`);

            return dbUtils.cleanup(res.pop());
        },
        async deleteAttribute(attrData: IAttribute): Promise<IAttribute> {
            // Delete links library<->attribute
            const libAttributesCollec = dbService.db.edgeCollection(LIB_ATTRIB_COLLECTION_NAME);

            // Delete all values
            await valueRepo.clearAllValues(attrData);

            const resDelEdges = await dbService.execute(aql`
                    FOR e IN ${libAttributesCollec}
                        FILTER e._to == ${'core_attributes/' + attrData.id}
                        REMOVE e IN ${libAttributesCollec}
                    `);

            // Delete attribute
            const col = dbService.db.collection(ATTRIB_COLLECTION_NAME);

            const res = await dbService.execute(aql`REMOVE ${{_key: attrData.id}} IN ${col} RETURN OLD`);

            // Return deleted attribute
            return dbUtils.cleanup(res.pop());
        },
        async getLibrariesUsingAttribute(attribute: IAttribute): Promise<ILibrary[]> {
            const libAttribCollec = dbService.db.edgeCollection(LIB_ATTRIB_COLLECTION_NAME);

            // TODO: use aql template tag, and find out why it doesn't work :)
            const query = `
                FOR v
                IN 1 INBOUND '${ATTRIB_COLLECTION_NAME}/${attribute.id}'
                ${LIB_ATTRIB_COLLECTION_NAME}
                RETURN v
            `;

            const res = await dbService.execute(query);

            return res.map(dbUtils.cleanup);
        }
    };
}
