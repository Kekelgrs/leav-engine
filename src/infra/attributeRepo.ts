import {IDbService} from 'infra/db/dbService';
import {IDbUtils} from 'infra/db/dbUtils';
import {IAttribute, AttributeTypes, AttributeFormats, IAttributeFilterOptions} from '../_types/attribute';
import {aql} from 'arangojs';
import {IValue} from '_types/value';
import {AqlQuery} from 'arangojs/lib/esm/aql-query';
import {ILibrary} from '_types/library';
import {ILibraryRepo, LIB_ATTRIB_COLLECTION_NAME} from './libraryRepo';
import {IValueRepo} from './valueRepo';

export interface IAttributeRepo {
    ATTRIB_COLLECTION_NAME?: string;
    getAttributes?(filters?: IAttributeFilterOptions): Promise<IAttribute[]>;
    updateAttribute?(attrData: IAttribute): Promise<IAttribute>;
    createAttribute?(attrData: IAttribute): Promise<IAttribute>;
    deleteAttribute?(attrData: IAttribute): Promise<IAttribute>;

    /**
     * Get all libraries for which this attribute is enabled
     *
     * @param attribute
     */
    getLibrariesUsingAttribute?(attribute: IAttribute): Promise<ILibrary[]>;
}

export const ATTRIB_COLLECTION_NAME = 'core_attributes';

export default function(
    dbService: IDbService | null = null,
    dbUtils: IDbUtils | null = null,
    libraryRepo: ILibraryRepo | null = null,
    valueRepo: IValueRepo | null = null
): IAttributeRepo {
    return {
        async getAttributes(filters?: IAttributeFilterOptions): Promise<IAttribute[]> {
            let query = `FOR a IN ${ATTRIB_COLLECTION_NAME}`;
            const bindVars = {};

            if (typeof filters !== 'undefined') {
                const dbFilters = dbUtils.convertToDoc(filters);
                const filtersKeys = Object.keys(dbFilters);
                for (let i = 0; i < filtersKeys.length; i++) {
                    query += ` FILTER a.@filterKey${i} == @filterValue${i}`;
                    bindVars[`filterKey${i}`] = filtersKeys[i];
                    bindVars[`filterValue${i}`] = `${dbFilters[filtersKeys[i]]}`;
                }
            }

            query += ` RETURN a`;

            const res = await dbService.execute({query, bindVars});

            return res.map(dbUtils.cleanup);
        },
        async updateAttribute(attrData: IAttribute): Promise<IAttribute> {
            const defaultParams = {
                _key: '',
                system: false,
                label: {fr: '', en: ''},
                type: AttributeTypes.ADVANCED,
                format: AttributeFormats.TEXT
            };
            let docToInsert = dbUtils.convertToDoc(attrData);

            docToInsert = {...defaultParams, ...docToInsert};

            // Insert in libraries collection
            const col = dbService.db.collection(ATTRIB_COLLECTION_NAME);
            const res = await dbService.execute(aql`UPDATE ${docToInsert} IN ${col} RETURN NEW`);

            return dbUtils.cleanup(res.pop());
        },
        async createAttribute(attrData: IAttribute): Promise<IAttribute> {
            const defaultParams = {
                _key: '',
                system: false,
                label: {fr: '', en: ''},
                type: AttributeTypes.ADVANCED,
                format: AttributeFormats.TEXT
            };
            let docToInsert = dbUtils.convertToDoc(attrData);

            docToInsert = {...defaultParams, ...docToInsert};

            // Insert in libraries collection
            const col = dbService.db.collection(ATTRIB_COLLECTION_NAME);
            const res = await dbService.execute(aql`INSERT ${docToInsert} IN ${col} RETURN NEW`);

            return dbUtils.cleanup(res.pop());
        },
        async deleteAttribute(attrData: IAttribute): Promise<IAttribute> {
            // Delete links library<->attribute
            const libAttributesCollec = dbService.db.edgeCollection(LIB_ATTRIB_COLLECTION_NAME);

            // Delete all values
            await valueRepo.clearAllValues(attrData);

            const resDelEdges = await dbService.execute(aql`
                    FOR e IN ${libAttributesCollec}
                        FILTER e._to == ${'core_attributes/' + attrData.id}
                        REMOVE e IN ${libAttributesCollec}
                    `);

            // Delete attribute
            const col = dbService.db.collection(ATTRIB_COLLECTION_NAME);
            const res = await dbService.execute(aql`REMOVE ${{_key: attrData.id}} IN ${col} RETURN OLD`);

            // Return deleted attribute
            return dbUtils.cleanup(res.pop());
        },
        async getLibrariesUsingAttribute(attribute: IAttribute): Promise<ILibrary[]> {
            const libAttribCollec = dbService.db.edgeCollection(LIB_ATTRIB_COLLECTION_NAME);

            // TODO: use aql template tag, and find out why it doesn't work :)
            const query = `
                FOR v
                IN 1 INBOUND '${ATTRIB_COLLECTION_NAME}/${attribute.id}'
                ${LIB_ATTRIB_COLLECTION_NAME}
                RETURN v
            `;

            const res = await dbService.execute(query);

            return res.map(dbUtils.cleanup);
        }
    };
}
