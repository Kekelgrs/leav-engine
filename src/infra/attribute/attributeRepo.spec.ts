import {Database} from 'arangojs';
import {IDbUtils} from 'infra/db/dbUtils';
import {AttributeFormats, AttributeTypes} from '../../_types/attribute';
import attributeRepo from '../attribute/attributeRepo';
import {IValueRepo} from '../value/valueRepo';

describe('AttributeRepo', () => {
    describe('getAttribute', () => {
        test('Get all attributes', async function() {
            const mockDbServ = {db: null, execute: global.__mockPromise([])};
            const mockDbUtils: Mockify<IDbUtils> = {
                findCoreEntity: global.__mockPromise([
                    {
                        id: 'label',
                        system: false,
                        label: {
                            fr: 'label'
                        }
                    }
                ])
            };

            const repo = attributeRepo(mockDbServ, mockDbUtils as IDbUtils);

            const trees = await repo.getAttributes();

            expect(mockDbUtils.findCoreEntity.mock.calls.length).toBe(1);
            expect(trees).toEqual([
                {
                    id: 'label',
                    system: false,
                    label: {
                        fr: 'label'
                    }
                }
            ]);
        });
    });

    describe('updateAttribute', () => {
        const docAttrData = {
            _key: 'test_attribute',
            system: true,
            format: 'text',
            type: 'standard',
            label: {fr: 'Test'}
        };

        const attrData = {
            id: 'test_attribute',
            system: true,
            label: {fr: 'Test'},
            format: AttributeFormats.TEXT,
            type: AttributeTypes.ADVANCED
        };

        test('Should update an existing attribute', async function() {
            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise([docAttrData])
            };

            const mockCleanupRes = attrData;
            const mockDbUtils = {
                cleanup: jest.fn().mockReturnValue(mockCleanupRes),
                convertToDoc: jest.fn().mockReturnValue(docAttrData)
            };

            const attrRepo = attributeRepo(mockDbServ, mockDbUtils);

            const updatedAttr = await attrRepo.updateAttribute(attrData);
            expect(mockDbServ.execute.mock.calls.length).toBe(1);
            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query).toMatch(/^UPDATE/);
            expect(mockDbServ.execute.mock.calls[0][0].query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].bindVars).toMatchSnapshot();

            expect(updatedAttr).toMatchObject(attrData);
        });
    });

    describe('createAttribute', () => {
        const docAttrData = {_key: 'test_attribute', system: true, format: 'text', type: 'standard'};
        const attrData = {
            id: 'test_attribute',
            system: true,
            label: {fr: 'Test'},
            format: AttributeFormats.TEXT,
            type: AttributeTypes.ADVANCED
        };

        test('Should create a new attribute', async function() {
            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise([docAttrData])
            };

            const mockCleanupRes = attrData;
            const mockDbUtils = {
                cleanup: jest.fn().mockReturnValue(mockCleanupRes),
                convertToDoc: jest.fn().mockReturnValue(docAttrData)
            };

            const attrRepo = attributeRepo(mockDbServ, mockDbUtils);

            const createdAttr = await attrRepo.createAttribute(attrData);
            expect(mockDbServ.execute.mock.calls.length).toBe(1);
            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query).toMatch(/^INSERT/);
            expect(mockDbServ.execute.mock.calls[0][0].query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].bindVars).toMatchSnapshot();

            expect(createdAttr).toMatchObject(attrData);
        });
    });

    describe('deleteAttribute', () => {
        const attrData = {
            id: 'test_attribute',
            system: false,
            label: {fr: 'Test'},
            format: AttributeFormats.TEXT,
            type: AttributeTypes.SIMPLE
        };

        const docAttrData = {
            _key: 'test_attribute',
            _id: 'core_attributes/test_attribute',
            _rev: '_WSgDYea--_',
            format: 'numeric',
            label: {en: 'Test', fr: 'Test'},
            system: false,
            type: 'index'
        };

        test('Should delete an attribute and return deleted attribute', async function() {
            const mockDbServ = {
                db: new Database(),
                execute: jest
                    .fn()
                    .mockReturnValueOnce([])
                    .mockReturnValueOnce(Promise.resolve([docAttrData]))
            };

            const mockCleanupRes = attrData;
            const mockDbUtils = {
                cleanup: jest.fn().mockReturnValue(attrData),
                convertToDoc: jest.fn().mockReturnValue(docAttrData)
            };

            const mockValueRepo: Mockify<IValueRepo> = {
                clearAllValues: jest.fn()
            };

            const attrRepo = attributeRepo(mockDbServ, mockDbUtils, mockValueRepo as IValueRepo);
            attrRepo.getAttributes = global.__mockPromise([attrData]);

            const deleteRes = await attrRepo.deleteAttribute(attrData);

            expect(mockDbServ.execute.mock.calls.length).toBe(2);

            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(typeof mockDbServ.execute.mock.calls[1][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].bindVars).toMatchSnapshot();

            expect(mockDbServ.execute.mock.calls[1][0].query).toMatch(/^REMOVE/);
            expect(mockDbServ.execute.mock.calls[1][0].query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[1][0].bindVars).toMatchSnapshot();
        });
    });
});
