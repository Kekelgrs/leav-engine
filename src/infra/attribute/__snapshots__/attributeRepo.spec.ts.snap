// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttributeRepo createAttribute Should create a new attribute 1`] = `"INSERT @value0 IN @@value1 RETURN NEW"`;

exports[`AttributeRepo createAttribute Should create a new attribute 2`] = `
Object {
  "@value1": "core_attributes",
  "value0": Object {
    "_key": "test_attribute",
    "format": "text",
    "label": Object {
      "fr": "Test",
    },
    "system": true,
    "type": "standard",
  },
}
`;

exports[`AttributeRepo deleteAttribute Should delete an attribute and return deleted attribute 1`] = `
"
                    FOR e IN @@value0
                        FILTER e._to == @value1
                        REMOVE e IN @@value0
                    "
`;

exports[`AttributeRepo deleteAttribute Should delete an attribute and return deleted attribute 2`] = `
Object {
  "@value0": "core_edge_libraries_attributes",
  "value1": "core_attributes/test_attribute",
}
`;

exports[`AttributeRepo deleteAttribute Should delete an attribute and return deleted attribute 3`] = `"REMOVE @value0 IN @@value1 RETURN OLD"`;

exports[`AttributeRepo deleteAttribute Should delete an attribute and return deleted attribute 4`] = `
Object {
  "@value1": "core_attributes",
  "value0": Object {
    "_key": "test_attribute",
  },
}
`;

exports[`AttributeRepo updateAttribute Should update an existing attribute 1`] = `"UPDATE @value0 IN @@value1 RETURN NEW"`;

exports[`AttributeRepo updateAttribute Should update an existing attribute 2`] = `
Object {
  "@value1": "core_attributes",
  "value0": Object {
    "_key": "test_attribute",
    "format": "text",
    "label": Object {
      "fr": "Test",
    },
    "multiple_values": false,
    "system": true,
    "type": "standard",
  },
}
`;
